### PR TITLE
add models for T2I-Adapter-XL

### PR DESCRIPTION
**Pull request by MC-E**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by MC-E** [comment]
We add T2I-Adapter-XL models to support the control on SDXL.

**Comment by sayakpaul** [comment]
@MC-E this looks good to me! Do we have any checkpoints that we can test this with?

File: src/diffusers/models/adapter.py [hunk 1]
```diff
@@ -165,6 +167,80 @@ def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
...
+                    self.body.append(
+                        AdapterResnetBlock_XL(channels[i - 1], channels[i], down=True))
+                elif (i == 1) and (j == 0):
+                    self.body.append(
+                        AdapterResnetBlock_XL(channels[i - 1], channels[i], down=False))
+                else:
+                    self.body.append(
+                        AdapterResnetBlock_XL(channels[i], channels[i], down=False))
+        self.body = nn.ModuleList(self.body)
+
```
Comment (patil-suraj) [inline src/diffusers/models/adapter.py:199]: Could we add some comments here explaining the logic?

File: src/diffusers/models/adapter.py [hunk 2]
```diff
@@ -165,6 +167,80 @@ def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
...
+        if self.downsample is not None:
+            x = self.downsample(x)
+        if self.in_conv is not None:
+            x = self.in_conv(x)
+        h = x
+        h = self.block1(h)
+        h = self.act(h)
+        h = self.block2(h)
+
+        return h + x
```
Comment (patil-suraj) [inline src/diffusers/models/adapter.py:241]: I think this can be supported in the `AdapterResnetBlock` with additional args. That might be better than adding a new block.

**Review by patil-suraj — COMMENTED** [review]
Very cool and minimal. just left two small comments.

File: src/diffusers/models/adapter.py (hunk 2 above)
Comment (MC-E) [reply src/diffusers/models/adapter.py]: @patil-suraj Thank you for this useful suggestion. I have made the modifications and implemented it using the existing AdapterResnetBlock.

Comment (MC-E) [reply src/diffusers/models/adapter.py]: It should be noted that when loading parameters, it is necessary to convert the state_dict to adapt to the existing pre-trained parameters:

state_dict_new = {}
    for key in state_dict:
        if 'in_conv' in key:
            idx_body = int(key[5])
            key_new = f'body.{idx_body//2}.in_conv.'+key.split('.')[-1]
        elif 'conv_in' in key:
            key_new = key
        else:
            idx_body = int(key[5])
            idx_body_new = idx_body//2
            idx_resnet_new = idx_body%2
            key_new = '.'.join(
                ['body',str(idx_body_new),'resnets',str(idx_resnet_new),key.split('.')[-2],key.split('.')[-1]]
            )
        state_dict_new[key_new] = state_dict[key]

Comment (sayakpaul) [reply src/diffusers/models/adapter.py]: Yes, that is true.

File: src/diffusers/models/adapter.py [hunk 3]
```diff
@@ -165,6 +167,48 @@ def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
         return features
 
 
+class FullAdapter_XL(nn.Module):
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:170]:
```suggestion
class FullAdapterXL(nn.Module): 
```

We tend to not use `_` in the class names.

Comment (MC-E) [reply src/diffusers/models/adapter.py:170]: Thanks, I will modify this part.

**Comment by sayakpaul** [comment]
@MC-E I tried loading an existing checkpoint here: https://colab.research.google.com/gist/sayakpaul/e9ed999df5714a6c9bb9c0a06cc9922a/scratchpad.ipynb. But I am currently facing issues. Could you look into it?

**Comment by MC-E** [comment]
> @MC-E I tried loading an existing checkpoint here: https://colab.research.google.com/gist/sayakpaul/e9ed999df5714a6c9bb9c0a06cc9922a/scratchpad.ipynb. But I am currently facing issues. Could you look into it?


It seems that all the key prefixes have an additional "adapter."?

**Comment by sayakpaul** [comment]
@MC-E I did the following:

```python
final_state_dict = {f"adapter.{k}": v for k, v in state_dict_new.items()}

xl_adapter = T2IAdapter(adapter_type="full_adapter_xl", downscale_factor=16)
xl_adapter.load_state_dict(final_state_dict)
```

It leads to:

```bash
RuntimeError: Error(s) in loading state_dict for T2IAdapter:
	size mismatch for adapter.conv_in.weight: copying a param with shape torch.Size([320, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([320, 768, 3, 3]).
```

Anything I am missing out on?

**Comment by MC-E** [comment]
Please follow the configs here https://github.com/TencentARC/T2I-Adapter/tree/XL/configs/inference; The number of input channels for OpenPose is three times that of others.

**Comment by sayakpaul** [comment]
I am using canny here:

```python
from huggingface_hub import hf_hub_download

repo_id = "TencentARC/T2I-Adapter"
filepath = "models_XL/adapter-xl-canny.pth"

state_dict_path = hf_hub_download(repo_id=repo_id, filename=filepath)
```

**Comment by MC-E** [comment]
@sayakpaul The input channels need to be 1. It seems that you set it as 3

**Comment by MC-E** [comment]
Happy to assist you. Feel free to ask any questions:)

**Comment by sayakpaul** [comment]
Yup, that fixed it :)

**Comment by sayakpaul** [comment]
@MC-E I created this conversion script for easy sharing and distribution: https://gist.github.com/sayakpaul/9a888466865bc0844d8979a2b2821c63. Let's add this maybe to the `scripts` directory. 

Also since you have already incorporated the feedback provided by @patil-suraj, I think for newer checkpoints (assuming we would use these blocks) we won't require any conversion.

**Comment by sayakpaul** [comment]
I think the next step would be to:

* Generate a `diffusers` variant of the checkpoints existing here: https://huggingface.co/TencentARC/T2I-Adapter/tree/main/models_XL. 
* Make changes to https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_adapter.py for SDXL-specific changes. 
* Run inference with the converted checkpoint (from pt. 1) to ensure everything is working as expected. 

Pt 2 mostly involves:

* Writing a new `encode_prompt()` method. We can readily use: https://github.com/huggingface/diffusers/blob/8d30d2579453a1a8f78c50fe57f7b8ba6ad8e36b/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L219. 
* Change how the `unet` is called following: https://github.com/huggingface/diffusers/blob/8d30d2579453a1a8f78c50fe57f7b8ba6ad8e36b/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L812.

We can add a new pipeline for this: `pipeline_stable_diffusion_xl_adapter.py`.

Let me know if you face any difficulties here. 

Then in a follow-up PR, we can add the training script to start our full-length experiments.

**Comment by MC-E** [comment]
@sayakpaul The XL pipeline is added. The hf models are uploaded at: https://huggingface.co/Adapter/t2iadapter/tree/main

**Comment by MC-E** [comment]
@sayakpaul The test code is:
```
import torch
from PIL import Image

from diffusers import (
    T2IAdapter,
    StableDiffusionXLAdapterPipeline,
    DDPMScheduler
)
from pytorch_lightning import seed_everything
from diffusers.models.unet_2d_condition import UNet2DConditionModel

sketch_image = Image.open('examples/edge_dog.png')

model_id = 'stabilityai/stable-diffusion-xl-base-1.0'

adapter = T2IAdapter.from_pretrained("Adapter/t2iadapter", subfolder='sketch_sdxl_1.0',torch_dtype=torch.float16, adapter_type="full_adapter_xl")
scheduler = DDPMScheduler.from_pretrained(model_id, subfolder="scheduler")

pipe = StableDiffusionXLAdapterPipeline.from_pretrained(
    model_id, adapter=adapter, safety_checker=None, torch_dtype=torch.float16, variant="fp16", scheduler=scheduler
)

pipe.to('cuda')
generator = torch.Generator().manual_seed(42)
sketch_image_out = pipe(prompt='a photo of a dog in real world, high quality', negative_prompt='extra digit, fewer digits, cropped, worst quality, low quality', image=sketch_image, generator=generator, guidance_scale=7.5).images[0]

sketch_image_out.save('sketch_image_out.png')

**Comment by sayakpaul** [comment]
@MC-E thanks for providing the code snippet and for your contributions! 

A couple of things:

* I simplified your snippet so that it's easier to follow. I hope that's okay. Could you please also leave a link to `edge_dog.png` so that we can try it out? 
* Do we need to ensure the inputs images of (1024, 1024) resolution as SDXL base model is typically good at that resolution?
* The default Euler Discrete scheduler of SDXL works well with a reduced `guidance_scale` (5.0 for example). So, I would try to reduce the `num_inference_steps` and `guidance_scale` and see the results.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 4]
```diff
@@ -0,0 +1,939 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:1]:
```suggestion
# Copyright 2023 TencentARC and The HuggingFace Team. All rights reserved.
```

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 5]
```diff
@@ -0,0 +1,939 @@
...
+    PIL_INTERPOLATION,
+    BaseOutput,
+    is_accelerate_available,
+    is_accelerate_version,
+    logging,
+    randn_tensor,
+    replace_example_docstring,
+)
+from ..pipeline_utils import DiffusionPipeline
+from basicsr.utils import tensor2img
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:45]: I think we can remove this no?

```suggestion
```

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 6]
```diff
@@ -0,0 +1,939 @@
...
+        model_sequence.extend([self.unet, self.vae, self.adapter])
+
+        hook = None
+        for cpu_offloaded_model in model_sequence:
+            _, hook = cpu_offload_with_hook(cpu_offloaded_model, device, prev_module_hook=hook)
+
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    def encode_prompt(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:246]: Let's add a similar `# Copied from ...` statement here in the comment?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 7]
```diff
@@ -0,0 +1,939 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:455]: `# Copied from` missing?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 8]
```diff
@@ -0,0 +1,939 @@
...
+        if latents is None:
+            latents = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            latents = latents.to(device)
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    def _get_add_time_ids(self, original_size, crops_coords_top_left, target_size, dtype):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:546]: `# Copied from ...` missing?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 9]
```diff
@@ -0,0 +1,939 @@
...
+            ),
+        )
+        # if xformers or torch_2_0 is used attention block does not need
+        # to be in float32 which can save lots of memory
+        if use_torch_2_0_or_xformers:
+            self.vae.post_quant_conv.to(dtype)
+            self.vae.decoder.conv_in.to(dtype)
+            self.vae.decoder.mid_block.to(dtype)
+
+    def _default_height_width(self, height, width, image):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:582]: `# Copied from ...` here as well? 

https://github.com/huggingface/diffusers/blob/052bf3280b9a3e83cdb06ab3a07645e23371e749/src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_adapter.py#L522

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 10]
```diff
@@ -0,0 +1,939 @@
...
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.disable_vae_tiling
+    def disable_vae_tiling(self):
+        r"""
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
+
+    def enable_model_cpu_offload(self, gpu_id=0):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:216]: `# Copied from ...` missing?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 11]
```diff
@@ -0,0 +1,939 @@
...
+        # height = height or self.default_sample_size * self.vae_scale_factor
+        # width = width or self.default_sample_size * self.vae_scale_factor
+
+        height, width = self._default_height_width(height, width, image)
+        device = self._execution_device
+
+        adapter_input = _preprocess_adapter_image(image, height, width).to(device)
+
+        original_size = (height, width) #original_size or (height, width)
+        target_size = (height, width) #target_size or (height, width)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:765]: Clean up the comments?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 12]
```diff
@@ -0,0 +1,939 @@
...
+
+        # here `guidance_scale` is defined analog to the guidance weight `w` of equation (2)
+        # of the Imagen paper: https://arxiv.org/pdf/2205.11487.pdf . `guidance_scale = 1`
+        # corresponds to doing no classifier free guidance.
+        do_classifier_free_guidance = guidance_scale > 1.0
+
+        # 3. Encode input prompt
+        # text_encoder_lora_scale = (
+        #     cross_attention_kwargs.get("scale", None) if cross_attention_kwargs is not None else None
+        # )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:800]: Clean up?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 13]
```diff
@@ -0,0 +1,939 @@
...
+        # make sure the VAE is in float32 mode, as it overflows in float16
+        if self.vae.dtype == torch.float16 and self.vae.config.force_upcast:
+            self.upcast_vae()
+            latents = latents.to(next(iter(self.vae.post_quant_conv.parameters())).dtype)
+
+        if not output_type == "latent":
+            image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False)[0]
+        else:
+            image = latents
+            return StableDiffusionAdapterXLPipelineOutput(images=image)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:928]: We use the `StableDiffusionXLPipelineOutput` here no?

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 5 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:45]: Yes, this needs to be deleted, as it is a residual from debugging.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 6 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:246]: Agreed, this part refers to the pipeline_stable_diffusion_xl.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 7 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:455]: Yes, also the pipeline_stable_diffusion_xl.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 8 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:546]: Yes, also the pipeline_stable_diffusion_xl.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 9 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:582]: Yes, also the pipeline_stable_diffusion_xl.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 10 above)
Comment (MC-E) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:216]: Yes, also the pipeline_stable_diffusion_xl.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 14]
```diff
@@ -0,0 +1,933 @@
...
+
+        >>> sketch_image = load_image('https://huggingface.co/Adapter/t2iadapter/resolve/main/sketch.png').convert('L')
+
+        >>> model_id = 'stabilityai/stable-diffusion-xl-base-1.0'
+
+        >>> adapter = T2IAdapter.from_pretrained("Adapter/t2iadapter", subfolder='sketch_sdxl_1.0',torch_dtype=torch.float16, adapter_type="full_adapter_xl")
+        >>> scheduler = DDPMScheduler.from_pretrained(model_id, subfolder="scheduler")
+
+        >>> pipe = StableDiffusionXLAdapterPipeline.from_pretrained(
+            model_id, adapter=adapter, safety_checker=None, torch_dtype=torch.float16, variant="fp16", scheduler=scheduler
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:65]: We don't need a safety checker as the base SDXL pipeline doesn't need it.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py (hunk 10 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:212]: You probably missed this one.

File: src/diffusers/pipelines/__init__.py [hunk 15]
```diff
@@ -117,6 +117,7 @@
         StableDiffusionXLPipeline,
     )
     from .t2i_adapter import StableDiffusionAdapterPipeline
+    from .t2i_adapter import StableDiffusionAdapterPipeline
```
Comment (sayakpaul) [inline src/diffusers/pipelines/__init__.py:120]: Needs to be modified to the XL pipeline.

Comment (MC-E) [reply src/diffusers/pipelines/__init__.py]: @sayakpaul Do you know the reason for this? It seems that I haven't modified the related files.
image

**Comment by sayakpaul** [comment]
> https://github.com/huggingface/diffusers/pull/4696#discussion_r1302555543

@MC-E I am looking into it.

**Comment by MC-E** [comment]
> > [#4696 (comment)](https://github.com/huggingface/diffusers/pull/4696#discussion_r1302555543)
> 
> @MC-E I am looking into it.


Thanks. Perhaps it is due to my modifications to models.unet_2d_condition.UNet2DConditionModel that caused the inconsistency in the copy?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py [hunk 16]
```diff
@@ -84,9 +84,9 @@ def get_dummy_components(self, adapter_type):
         torch.manual_seed(0)
         adapter = T2IAdapter(
             in_channels=3,
-            channels=[32, 64],
+            channels=[320, 640, 1280, 1280],
             num_res_blocks=2,
-            downscale_factor=2,
+            downscale_factor=8,
```
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:89]: Can't we reduce this for testing purposes like the way they previously were?

Comment (sayakpaul) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:89]: The UNet also has the same number of channels in the tests.

File: src/diffusers/models/adapter.py [hunk 17]
```diff
@@ -260,7 +260,7 @@ class LightAdapter(nn.Module):
     def __init__(
         self,
         in_channels: int = 3,
-        channels: List[int] = [320, 640, 1280],
+        channels: List[int] = [320, 640, 1280, 1280],
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:263]: Why is this change needed?

**Comment by sayakpaul** [comment]
@MC-E I tried running inference however, there seems to be an error. Here's my Colab:
https://colab.research.google.com/gist/sayakpaul/8410f72f5bbf5535c76f78bed5d166c2/scratchpad.ipynb

```bash
RuntimeError: Given groups=1, weight of size [320, 256, 3, 3], expected input[1, 768, 64, 64] to have 256 channels, but got 768 channels instead
```

(Full error in the Colab Notebook)

**Comment by sayakpaul** [comment]
Indeed the results seem to be affected by the choice of scheduler. 

As @MC-E mentioned via Slack, since the concerned checkpoints were trained with the DDPM scheduler, it tends to rely on it. 

For example, with DDPM, the results looks quite nice:

image


However, with Euler Discrete (the default for SDXL), the results are worse:

image

Here's my Colab Notebook in case anyone wants to play with this.

File: src/diffusers/models/adapter.py [hunk 18]
```diff
@@ -152,6 +154,48 @@ def __init__(
...
+        x = self.unshuffle(x)
+        x = self.conv_in(x)
+
+        features = []
+
+        for block in self.body:
+            x = block(x)
+            features.append(x)
+        
+        return features
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:167]: Not sure why it's shown as a change as it's same with `main`

https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/adapter.py#L155C1-L165C24

File: src/diffusers/models/adapter.py [hunk 19]
```diff
@@ -221,6 +265,7 @@ def __init__(
         downscale_factor: int = 8,
     ):
         super().__init__()
+        print(f"From {self.__class__} channels: {channels}.")
```
Comment (sayakpaul) [inline src/diffusers/models/adapter.py:268]:
```suggestion
```

Sorry I mistakenly introduced it.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py [hunk 20]
```diff
@@ -165,11 +172,8 @@ def test_stable_diffusion_adapter_default_case(self):
 
         inputs = self.get_dummy_inputs(device)
         image = sd_pipe(**inputs).images
-        image_slice = image[0, -3:, -3:, -1]
 
         assert image.shape == (1, 64, 64, 3)
-        expected_slice = np.array([0.4965, 0.5548, 0.4330, 0.4771, 0.6226, 0.4382, 0.5037, 0.5071, 0.4782])
-        assert np.abs(image_slice.flatten() - expected_slice).max() < 5e-3
```
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:172]: Same as https://github.com/huggingface/diffusers/pull/4696/files#r1302892199.

Comment (MC-E) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:172]: Okey, 
What is the function of these lines of code?

Comment (sayakpaul) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:172]: They check for the expected outputs.

Comment (MC-E) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_adapter.py:172]: Okey, get it.

**Comment by sayakpaul** [comment]
Yes, sure.

**Comment by sayakpaul** [comment]
Thanks for the changes, @MC-E!

Let's maybe also add a test for the Adapter XL pipeline here? Then I think we can merge. @patrickvonplaten could you also give this a look?

**Comment by MC-E** [comment]
> Thanks for the changes, @MC-E!
> 
> Let's maybe also add a test for the Adapter XL pipeline here? Then I think we can merge. @patrickvonplaten could you also give this a look?

@sayakpaul Okey, I will write a test for XL pipeline.

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 21]
```diff
@@ -0,0 +1,944 @@
...
+        ...     negative_prompt="extra digit, fewer digits, cropped, worst quality, low quality",
+        ...     image=sketch_image,
+        ...     generator=generator,
+        ...     guidance_scale=7.5,
+        ... ).images[0]
+        ```
+"""
+
+
+def _preprocess_adapter_image(image, height, width):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:83]: In a follow-up we can leverage the image processor module here and in the SD T2I adapter pipeline.

**Review by sayakpaul — APPROVED** [review]
Looking good to me. Thanks a lot for iterating.

**Comment by sayakpaul** [comment]
Failing tests seem unrelated.

**Comment by MC-E** [comment]
> Failing tests seem unrelated.


mean? Are there any issues that need to be further resolved?

**Comment by sayakpaul** [comment]
Does not seem like so. I have requested a review from @patrickvonplaten who's the core maintainer of diffusers.

**Comment by MC-E** [comment]
> Does not seem like so. I have requested a review from @patrickvonplaten who's the core maintainer of diffusers.

Okey, thanks!

**Comment by MC-E** [comment]
@patrickvonplaten Could you help us check it? There is an error unrelated to this PR. It was not there before the last modification. The last modification just sorted the order of the imports: https://github.com/huggingface/diffusers/pull/4696/commits/4194ad1b00c8c8af1fad14e29ab3a01a75d7c738

**Review by patrickvonplaten — APPROVED** [review]
Very clean! Nice work. 

Can you also add this class to the docs here: https://github.com/huggingface/diffusers/blob/7e5587a5ac07be5621bfbbe24868b79e1cd0aede/docs/source/en/api/pipelines/stable_diffusion/adapter.md?plain=1#L178

and then maybe also add a couple of examples there?

**Comment by sayakpaul** [comment]
Should be added to a separate doc page for Adapter XL for now. @stevhliu will then do a larger effort to unify pipelines that have both SD and SDXL variants.

**Comment by sayakpaul** [comment]
@MC-E could you please run `make style && make quality`? @patrickvonplaten WDYT about the docs?

**Comment by MC-E** [comment]
> @MC-E could you please run `make style && make quality`? @patrickvonplaten WDYT about the docs?
@sayakpaul Yesh, I can run it. The style and quality at this time seem to have no issues?

**Comment by sayakpaul** [comment]
The usage examples use single quotes and we're all about double quotes in `diffusers` :-)

**Comment by sayakpaul** [comment]
> Are there any other issues in docs? Let me fix them together.

Nope, looks good to me.

**Comment by MC-E** [comment]
Okay:)

File: src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py [hunk 22]
```diff
@@ -0,0 +1,944 @@
...
+                # compute the previous noisy sample x_t -> x_t-1
+                latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
+
+                # call the callback, if provided
+                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
+
+        # make sure the VAE is in float32 mode, as it overflows in float16
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/t2i_adapter/pipeline_stable_diffusion_xl_adapter.py:924]: Can we update this part of the code to conform with the changes here: https://github.com/huggingface/diffusers/pull/4796 ?

**Review by patrickvonplaten — COMMENTED** [review]
Looks very nice!

Could we maybe add one slow tests (or should we wait until some powerful checkpoints are out?)

Apart from this left one nit & overall looks good to  merge for me!

**Comment by MC-E** [comment]
@patrickvonplaten We have updated the code to conform with the changes in https://github.com/huggingface/diffusers/pull/4796 

https://github.com/huggingface/diffusers/pull/4696/commits/08780772ddc155bd341e7ccdb86546a3caf42753

**Comment by sayakpaul** [comment]
@MC-E we need to resolve the conflicts.

**Comment by sayakpaul** [comment]
@MC-E once the conflicts are addressed, I am happy to merge this!

**Comment by MC-E** [comment]
Okay, on it:)

**Comment by sayakpaul** [comment]
@MC-E I think you accidentally borked the commit history. It shouldn't be like that.

**Comment by MC-E** [comment]
In order to resolve conflicts, I perform a rebase on this PR:(

**Comment by sayakpaul** [comment]
Thanks for your amazing contribution!

**Comment by MC-E** [comment]
Thank you for your help:)

**Comment by sayakpaul** [comment]
The documentation is updated: https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/adapter#usage-example-with-the-base-model-of-stablediffusionxl.

**Comment by MC-E** [comment]
Okay, thanks:)
